doc/shared/sunadjoint/SUNAdjointStepper.rst

```diff
@@ -0,0 +1,247 @@
+..
+   ----------------------------------------------------------------
+   SUNDIALS Copyright Start
+   Copyright (c) 2002-2024, Lawrence Livermore National Security
+   and Southern Methodist University.
+   All rights reserved.
+
+   See the top-level LICENSE and NOTICE files for details.
+
+   SPDX-License-Identifier: BSD-3-Clause
+   SUNDIALS Copyright End
+   ----------------------------------------------------------------
+
+.. _SUNAdjointStepper:
+
+The SUNAdjointStepper API and Module
+====================================
+
+The :c:type:`SUNAdjointStepper` API and module provides a package-agnostic
+interface to SUNDIALS ASA capabilities. It currently only supports the discrete
+ASA capabilities in the ARKODE package, but in the future this support may be expanded.
+
+A :c:type:`SUNAdjointStepper` is a pointer to the
+:c:struct:`SUNAdjointStepper_` structure:
+
+.. c:type:: struct SUNAdjointStepper_ *SUNAdjointStepper
+
+.. c:struct:: SUNAdjointStepper_
+
+   .. c:member:: SUNStepper adj_sunstepper
+
+      The :c:type:`SUNStepper` object used for backwards time stepping of the adjoint ODE system.
+
+   .. c:member:: SUNStepper fwd_sunstepper
+
+      The :c:type:`SUNStepper` object used for forward time stepping of the original ODE system if any recomputation of missing
+      state data is required during the backwards integration.
+
+   .. c:member:: sunrealtype tf
+
+      The terminal time of the backwards adjoint ODE.
+
+   .. c:member:: int64_t step_idx
+
+      The index of the current backward integration step with respect to the forward integration.
+
+   .. c:member:: int64_t final_step_idx
+
+      The index of the final step in the forward integration (corresponds to ``tf``).
+
+   .. c:member:: SUNMatrix Jac
+
+      Matrix data for the Jacobian :math:`df/dy`.
+
+   .. c:member:: SUNMatrix JacP
+
+      Matrix data for the Jacobian :math:`df/dp`.
+
+   .. c:member:: SUNJacFn JacFn
+
+      Jacobian function pointer to evaluate :math:`df/dy`.
+
+   .. c:member:: SUNJacFn JacPFn
+
+      Jacobian function pointer to evaluate :math:`df/dp`.
+
+   .. c:member:: SUNJacTimesFn JvpFn
+
+      Jacobian-times-vector function pointer to evaluate :math:`(df/dy)^T v`.
+
+   .. c:member:: SUNJacTimesFn JPvpFn
+
+      Jacobian-times-vector function pointer to evaluate :math:`(df/dp)^T v`.
```

Comment:
I guess I'm confused at the different between `JPvpFn` which is documented as `(df/dp)^T v` versus `vJPpFn` which is `v^T(df/dp)`. Mathematically these two quantities are just transposes of each other with the first a column vector and the second a row vector. N_Vector doesn't distinguish between the two, so I don't see the difference.

Reply:
The former corresponds to forward-mode automatic differentiation and latter reverse-mode, which why I setup both. However, you are right that from the integrator's perspective it does not matter since both give effectively the same result. So we could probably eliminate one of these function pointers and just let a user decide if its implemented as `(df/dp)^T v` or `v^T(df/dp)`.